### PR TITLE
Support multiple keys and use new webauthn endpoints

### PR DESCRIPTION
**Pull request by hobbitronics**
This picks up where https://github.com/silinternational/idp-profile-ui/pull/105 left off, which I closed without merging.

~~We are still trying to get 'change' and 'remove' endpoints working properly.~~

image
Updated now has created dates and editing labels works even when only 1 key!
image

### CODE REVIEW

**Comment by Baggerone** [comment]
(Copied from our slack channel)

Maybe you haven't gotten to this yet, but when I try to update a webauthn child, the UI does a PUT to `https://profile-api.gtis.guru/mfa/2` but doesn't include the `1/webauthn/` in the middle of it.  (I.e. the parent id before webauthn)

Same problem for deleting a webauthn child.

At this point, (as we kind of mentioned already), you're including a card for the parent mfa record in addition to all the children webauthns .  You should only show the children.

**Comment by hobbitronics** [comment]
> (Copied from our slack channel)
> 
> Maybe you haven't gotten to this yet, but when I try to update a webauthn child, the UI does a PUT to `https://profile-api.gtis.guru/mfa/2` but doesn't include the `1/webauthn/` in the middle of it. (I.e. the parent id before webauthn)
> 
> Same problem for deleting a webauthn child.
> 
> At this point, (as we kind of mentioned already), you're including a card for the parent mfa record in addition to all the children webauthns . You should only show the children. (In other words, the parent mfa doesn't hold any information that the user needs to see. It's not a webauthn record itself.)

If we want to show the dates then we will need to show that in the "master" or "summary" card which will not be representing a key itself (@Schparky does that sound correct).

I still haven't decided upon a layout that I really like, but I don't think we should get terribly hung up on that either.

**Comment by Schparky** [comment]
> If we want to show the dates then we will need to show that in the "master" or "summary" card which will not be representing a key itself (@Schparky does that sound correct).

I'm not sure what dates you're referring to.

**Comment by hobbitronics** [comment]
> > If we want to show the dates then we will need to show that in the "master" or "summary" card which will not be representing a key itself (@Schparky does that sound correct).
> 
> I'm not sure what dates you're referring to.

The Created and Last Used dates.

**Comment by Schparky** [comment]
> > > If we want to show the dates then we will need to show that in the "master" or "summary" card which will not be representing a key itself (@Schparky does that sound correct).
> > 
> > 
> > I'm not sure what dates you're referring to.
> 
> The Created and Last Used dates.

Is that because the "child" cards don't have those dates?

**Comment by hobbitronics** [comment]
> > > > If we want to show the dates then we will need to show that in the "master" or "summary" card which will not be representing a key itself (@Schparky does that sound correct).
> > > 
> > > 
> > > I'm not sure what dates you're referring to.
> > 
> > 
> > The Created and Last Used dates.
> 
> Is that because the "child" cards don't have those dates?

Exactly. They only have "id" and "label"

**Comment by Baggerone** [comment]
> > Is that because the "child" cards don't have those dates?
> 
> Exactly. They only have "id" and "label"

They have those dates in the database. So, I guess I need to add them to the return object, assuming they're being set.  I'll check on that.

File: src/2sv/Change.vue
Comment (Baggerone) [inline src/2sv/Change.vue:43]: This doesn't look right to me.
Is the order of the parameters right?
Is the hard coded `0` index right?

File: src/profile/MfaCardLabel.vue
Comment (Baggerone) [inline src/profile/MfaCardLabel.vue:69]:
```suggestion
      const mfa = await changeWebauthn(this.id, this.keyId, {
```

File: src/profile/MfaCardLabel.vue
Comment (Baggerone) [inline src/profile/MfaCardLabel.vue:48]: Does this get hydrated properly so that it doesn't end up as `undefined` down in the call to `changeWebauthn()`?

File: src/2sv/Change.vue
Comment (hobbitronics) [reply src/2sv/Change.vue:43]: The order does look incorrect here, but if the user will only "change" keys if they only have 1 left so that's why I hard coded it.

File: src/global/mfa.js
Comment (forevermatt) [inline src/global/mfa.js:27]: I think [`find()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find) only returns the first matching record, not all of the matching records (like `filter()` does). Is that the desired behavior?

File: src/profile/Index.vue
Comment (forevermatt) [inline src/profile/Index.vue:104]: What is the data structure here? I expected `keys` to be an array of webauthn keys, but is that actually down in `data`?

**Review by forevermatt — COMMENTED** [review]
I have a few questions, but bigger picture, I pulled Alex into a quick discussion (our Friday afternoon) about what would probably work well for this page now that we want to support potentially multiple security keys. (The lack of a clarity about the goal looked like it was hindering this work, and I wanted to get that roadblock out of the way, both to make this work easier and reduce later rework that would be needed.)

The result of that was some specific things to aim for (I have a Google Doc I can share with you), but I don't know whether it's better to get this merged first (since it sounds like it's working), or for you to go ahead and do the additional desired changes here before merging. Thoughts?

**Review by Baggerone — APPROVED** [review]
By approving this, I'm just saying that it appears to be working for me locally.  I'll leave review of the javascript to Matt.

**Comment by Baggerone** [comment]
I guess we should coordinate merging of this PR with merging the idp-id-broker PR.

File: src/global/mfa.js
Comment (forevermatt) [reply src/global/mfa.js:27]: @hobbitronics Just following up, is `find()` really the appropriate method to use here? What kind of data is `mfa.keys` (aka. Object vs Array)?

**Comment by forevermatt** [comment]
It looks like we should change the "Lost or New" button to simply a "Remove" button when the user has a single Security Key, since the user no longer needs to remove an existing security key when they get a new one.

**Comment by forevermatt** [comment]
I think replacing that "Lost or New" button with a "Remove" button is the only thing that needs to be changed before I'm fine with approving this PR.

File: src/global/mfa.js
Comment (forevermatt) [reply src/global/mfa.js:27]: Okay, I see that mfa.keys is an object, not an array:
```json
{
  "totp": {},
  "u2f": {},
  "keys": {
    "id": 2,
    "type": "webauthn",
    "label": "Security Key",
    "created_utc": "2022-12-15T19:42:12Z",
    "last_used_utc": "2022-12-15T19:50:21Z",
    "data": [
      {
        "id": 3,
        "label": "SIL Mac TouchID",
        "last_used_utc": null,
        "created_utc": "2022-12-15 19:42:15"
      },
      {
        "id": 4,
        "label": "Blue Yubikey2",
        "last_used_utc": null,
        "created_utc": "2022-12-15 19:42:37"
      },
      {
        "id": 5,
        "label": "Personal smartphone",
        "last_used_utc": null,
        "created_utc": "2022-12-15 19:43:44"
      }
    ]
  },
  "backup": {},
  "numVerified": 3
}
```

Given that, it seems like `keys` is a bit misleading, if we want to change that back to `webauthn` in a future PR. Just fwiw.

**Review by forevermatt — APPROVED** [review]

**Comment by forevermatt** [comment]
@hobbitronics Is this ready to merge?

**Comment by hobbitronics** [comment]
> @hobbitronics Is this ready to merge?

Yes, but will want to wait for the api and broker to be merged at the same time.

**Pull request merged**
